--- a/port/cpl_curl_priv.h
+++ b/port/cpl_curl_priv.h
@@ -31,6 +31,14 @@
 
 #include <curl/curl.h>
 
+/*
+ To maintain compatibility with macOS 10.15(.5) which ships with curl 7.64.1, we redefine
+ LIBCURL_VERSION_NUM from the SDK's curlver.h to that version, so gdal will not try to 
+ access newer APIs like curl_multi_poll() appearing in 7.66.0 used in cpl_http.cpp.
+ */
+#undef LIBCURL_VERSION_NUM
+#define LIBCURL_VERSION_NUM 0x074001
+
 #ifndef CURL_AT_LEAST_VERSION
 #define CURL_VERSION_BITS(x,y,z) ((x)<<16|(y)<<8|z)
 #define CURL_AT_LEAST_VERSION(x,y,z) \
