--- a/nctest/CMakeLists.txt
+++ b/nctest/CMakeLists.txt
@@ -18,7 +18,9 @@
 ADD_TEST(nctest ${EXECUTABLE_OUTPUT_PATH}/nctest)
 
 add_bin_test_no_prefix(tst_rename)
-add_sh_test(nctest compare_test_files)
-IF(HAVE_BASH)
-  SET_TESTS_PROPERTIES(nctest_compare_test_files PROPERTIES DEPENDS nctest)
-ENDIF(HAVE_BASH)
+IF(BUILD_UTILITIES)
+  add_sh_test(nctest compare_test_files)
+  IF(HAVE_BASH)
+    SET_TESTS_PROPERTIES(nctest_compare_test_files PROPERTIES DEPENDS nctest)
+  ENDIF(HAVE_BASH)
+ENDIF(BUILD_UTILITIES)
