diff --git a/src/hnswvacuum.c b/src/hnswvacuum.c
index 251d9d9ad..2f7b2f372 100644
--- a/src/hnswvacuum.c
+++ b/src/hnswvacuum.c
@@ -9,6 +9,10 @@
 #include "storage/lmgr.h"
 #include "utils/memutils.h"
 
+#if PG_VERSION_NUM >= 180000
+#define vacuum_delay_point() vacuum_delay_point(false)
+#endif
+
 /*
  * Check if deleted list contains an index TID
  */
diff --git a/src/ivfvacuum.c b/src/ivfvacuum.c
index 1272da831..750cdd65a 100644
--- a/src/ivfvacuum.c
+++ b/src/ivfvacuum.c
@@ -5,6 +5,10 @@
 #include "ivfflat.h"
 #include "storage/bufmgr.h"
 
+#if PG_VERSION_NUM >= 180000
+#define vacuum_delay_point() vacuum_delay_point(false)
+#endif
+
 /*
  * Bulk delete tuples from the index
  */
