--- a/topology/Makefile.in
+++ b/topology/Makefile.in
@@ -65,6 +65,11 @@
 # Extra files to remove during 'make clean'
 EXTRA_CLEAN=$(SQL_OBJS) topology_upgrade.sql.in
 
+# Set this before the include of PGXS
+top_builddir = @top_builddir@
+INSTALL = @INSTALL@
+LIBTOOL := @LIBTOOL@
+
 # PGXS information
 PG_CONFIG := @PG_CONFIG@
 PGXS := @PGXS@
@@ -73,10 +78,7 @@
 NO_TEMP_INSTALL=yes
 include $(PGXS)
 
-top_builddir = @top_builddir@
 VPATH := @srcdir@
-INSTALL = @INSTALL@
-LIBTOOL = @LIBTOOL@
 
 # Set PERL _after_ the include of PGXS
 PERL=@PERL@
