--- a/src/backend/replication/basebackup.c
+++ b/src/backend/replication/basebackup.c
@@ -240,6 +240,7 @@
 	{"pg_filenode.map", false},
 	{"pg_internal.init", true},
 	{"PG_VERSION", false},
+	{".DS_Store", false},
 #ifdef EXEC_BACKEND
 	{"config_exec_params", true},
 #endif
--- a/src/bin/pg_verify_checksums/pg_verify_checksums.c
+++ b/src/bin/pg_verify_checksums/pg_verify_checksums.c
@@ -73,6 +73,7 @@
 	{"pg_filenode.map", false},
 	{"pg_internal.init", true},
 	{"PG_VERSION", false},
+	{".DS_Store", false},
 #ifdef EXEC_BACKEND
 	{"config_exec_params", true},
 #endif
