--- a/Makefile
+++ b/Makefile
@@ -8,7 +8,7 @@
 INCLUDEDIR_SERVER := ${shell $(PG_CONFIG) --includedir-server}
 
 CP = cp
-PG_CFLAGS := -fPIC -Wall -Wextra -Wno-unused-parameter -Wno-declaration-after-statement -Wno-cast-function-type -std=c11 -DPLJS_VERSION=\"$(PLJS_VERSION)\" -g -O0
+PG_CFLAGS := -fPIC -Wall -Wextra -Wno-unused-parameter -Wno-declaration-after-statement -Wno-cast-function-type -Wno-int-conversion -Wno-implicit-function-declaration -std=c11 -DPLJS_VERSION=\"$(PLJS_VERSION)\" -g -O0
 SRCS = src/pljs.c src/cache.c src/functions.c src/types.c src/params.c
 OBJS = src/pljs.o src/cache.o src/functions.o src/types.o src/params.o deps/quickjs/libquickjs.a
 MODULE_big = pljs
