diff --git a/CMakeLists.txt b/CMakeLists.txt
index 224f4b2adfe99b668cbcf20738bb57c0bce5d4f2..0a995140d157d35d6f96c9dc4bc84e467af81c0a 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -133,17 +133,13 @@ else()
 	message( STATUS "Boost_USE_MULTITHREAD=OFF" )
 endif()
 
-#-- minimalist build allowed with boost version older than 1.48
-set( SFCGAL_Boost_COMPONENTS thread system serialization )
+#-- minimalist build allowed with boost version 1.69 or newer
+set( SFCGAL_Boost_COMPONENTS thread serialization )
 #-- program_options
 if ( SFCGAL_BUILD_TESTS OR SFCGAL_BUILD_EXAMPLES OR SFCGAL_BUILD_OSG )
 	set( SFCGAL_Boost_COMPONENTS chrono unit_test_framework filesystem program_options timer ${SFCGAL_Boost_COMPONENTS} )
 endif()
-find_package( Boost COMPONENTS ${SFCGAL_Boost_COMPONENTS} REQUIRED )
-if((${Boost_MAJOR_VERSION} EQUAL 1) AND (${Boost_MINOR_VERSION} EQUAL 58))
-	message( STATUS "Defining BOOST_VARIANT_USE_RELAXED_GET_BY_DEFAULT" )
-	add_definitions( "-DBOOST_VARIANT_USE_RELAXED_GET_BY_DEFAULT" )
-endif()
+find_package( Boost 1.69.0 COMPONENTS ${SFCGAL_Boost_COMPONENTS} REQUIRED )
 
 if( SFCGAL_USE_STATIC_LIBS )
   add_definitions( "-DSFCGAL_USE_STATIC_LIBS" )
