--- a/ogr/ogrsf_frmts/sqlite/GNUmakefile
+++ b/ogr/ogrsf_frmts/sqlite/GNUmakefile
@@ -35,8 +35,7 @@
 CPPFLAGS +=  -DHAVE_RASTERLITE2 $(RASTERLITE2_CFLAGS)
 endif
 
-# We may lie by defining SQLITE_HAS_NON_DEPRECATED_AUTO_EXTENSION with MacOS SQLite3. The correct check is done for CMake builds
-CPPFLAGS	:=	-iquote ..  $(SQLITE_INC) $(SPATIALITE_INC) $(PROJ_INCLUDE) $(PROJ_FLAGS) $(CPPFLAGS) -DSQLITE_HAS_NON_DEPRECATED_AUTO_EXTENSION -DOGR_SQLITE_ALLOW_LOAD_EXTENSIONS
+CPPFLAGS	:=	-iquote ..  $(SQLITE_INC) $(SPATIALITE_INC) $(PROJ_INCLUDE) $(PROJ_FLAGS) $(CPPFLAGS)
 
 default:	$(O_OBJ:.o=.$(OBJ_EXT))
 
