--- a/port/cpl_curl_priv.h
+++ b/port/cpl_curl_priv.h
@@ -31,6 +31,14 @@
 
 #include <curl/curl.h>
 
+/*
+ To maintain compatibility with macOS <= 10.14 which ships with curl 7.54.0, we redefine
+ LIBCURL_VERSION_NUM from the SDK's curlver.h to that version, so gdal will not try to 
+ access newer APIs like curl_mime_addpart appearing in 7.56.0
+ */
+#undef LIBCURL_VERSION_NUM
+#define LIBCURL_VERSION_NUM 0x073600
+
 #ifndef CURL_AT_LEAST_VERSION
 #define CURL_VERSION_BITS(x,y,z) ((x)<<16|(y)<<8|z)
 #define CURL_AT_LEAST_VERSION(x,y,z) \
