--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -281,7 +281,7 @@
 
 install(TARGETS geos geos_cxx_flags
   EXPORT geos-targets
-  LIBRARY DESTINATION lib NAMELINK_SKIP
+  LIBRARY DESTINATION lib
   ARCHIVE DESTINATION lib
   RUNTIME DESTINATION bin
   INCLUDES DESTINATION include
@@ -318,6 +318,8 @@
 endif()
 install(FILES "${CMAKE_CURRENT_BINARY_DIR}/capi/geos_c.h"
   DESTINATION include)
+install(FILES "${CMAKE_CURRENT_LIST_DIR}/include/geos.h"
+  DESTINATION include)
 
 add_subdirectory(tools)
 add_subdirectory(util)
